libqtile/utils.py

```diff
@@ -209,3 +211,39 @@ def send_notification(title, message, urgent=False, timeout=10000):
         notifier.show()
     except Exception as exception:
         logger.error(exception)
+
+
+def guess_terminal():
+    """Try to guess terminal."""
+    test_terminals = [
+            ['x-terminal-emulator', '-?'],
+            ['roxterm', '-?'],
+            ['sakura', '-v'],
+            ['hyper', '-v'],
+            ['alacritty', '-V'],
+            ['terminator', '-v'],
+            ['gnome-terminal', '-?'],
+            ['konsole', '--help'],
+            ['xfce4-terminal', '-V'],
+            ['lxterminal', '-v'],
+            ['mate-terminal', '-h'],
+            ['tilda', '--help'],
+            ['guake', '-h'],
+            ['eterm', '-h'],
+            ['st', '-v'],
+            ['urxvt', '-help'],
+            ['xterm', '-help'],
+            ]
+
+    to_null = ['>', '/dev/null']
+    for terminal in test_terminals:
+        logger.debug('Guessing terminal: {}'.format(terminal[0]))
+        try:
+            call(terminal + to_null)
```

Comment:
`subprocess.run()` seems like it'd be useful here. Also, adding a timeout is probably a good idea.

Reply:
What about ``subprocess.check_call`` on ``which <terminal name>``?
This way one doesn't need to maintain a list of options to call the terminal emulators with immediate return and there's no need for a timeout. Or is ``which`` not portable enough?

Reply:
> `subprocess.run()` seems like it'd be useful here. Also, adding a timeout is probably a good idea.

I missmatched python versions and did not noted `run`. But why timeout is needed?

Reply:
> What about `subprocess.check_call` on `which <terminal name>`?
> This way one doesn't need to maintain a list of options to call the terminal emulators with immediate return and there's no need for a timeout. Or is `which` not portable enough?

I tried that but the flow didn't worked. Speaking the true, I used python 2 documentation, so maybe something changed

Reply:
Sorry, my bad, ``subprocess.check_call`` was superseeded by ``subprocess.run`` in Python 3.5.

However:
  - the proper way to redirect standard output to the null data sink is passing ``stdout=subprocess.DEVNULL`` to ``subprocess.run``, not a shell redirection without ``shell=True`` (the call ``subprocess.run(["xterm", "-help", ">", "/dev/null"])`` indeed succeeds with return code 0 when ``xterm`` is installed on the system but the redirection is not effective and ``xterm`` complains with `xterm: No absolute path found for shell: >`). Standard error should also be redirected to ``/dev/null`` since for example ``st -v`` writes on standard error (and exits with status 1).

  - Agreed that it is unlikely that terminal emulators change how to access their help or version number. I find it nevertheless useless to maintain a list of these options together with a list of terminal emulators. Calling ``which`` has the following advantages:
    + no need to check/maintain which options need to be passed to the terminal emulators to silently check their presence;
    + consistent interface: if ``term`` is there, ``which term`` exits with 0, else with 1;
    + no call to the actual terminal emulator, no real need for a timeout (although something like 1 second would still be reasonable);
    + ``subprocess.run(["which", "<terminal>"], check=True, stdout=DEVNULL)`` in a try-except block catching ``CalledProcessError`` is clean and explicit. Catching ``FileNotFoundError`` means that ``which`` is not there and the search can stop immediately.

``which`` seems [widespread enough](https://en.wikipedia.org/wiki/Which_(command)) to me but I would not argue if someone comes with a decent counter-example

IMHO, the first point should be fixed. I'm happily discussing the second one further if wanted.

Reply:
@airwoodix thanks for such good explaination :) Fixed as you suggested

Reply:
If using `which`, what about [`shutil.which()`](https://docs.python.org/3/library/shutil.html#shutil.which) then?

Reply:
:D @pawamoy good command, too. But maybe let's stop here

Reply:
@pawamoy good point, of course I forgot to check ``shutil``... I also think it's a better match than the subprocess call. It removes the dependency to ``which`` and the interface is even cleaner. 
@zordsdavini it doesn't cost much to fix that, right?

Reply:
6 x 5€ = 30€

Reply:
pushed

Reply:
thanks